lib/socketpair.c

```diff
@@ -94,18 +96,31 @@ int Curl_socketpair(int domain, int type, int protocol,
     goto error;
   if(connect(socks[0], &a.addr, sizeof(a.inaddr)) == -1)
     goto error;
+
+  /* use non-blocking accept to make sure we don't block forever */
+  if(ioctlsocket(listener, FIONBIO, &nonblock) == -1)
+    goto error;
+  FD_ZERO(&fds);
+  FD_SET(listener, &fds);
+  /* this is Windows-only code -> using select is fine regardless
+     of the _value_ of the socket descriptor, also nfds is ignored */
+  select(0, &fds, NULL, NULL, &timeout);
```

Comment:
Then shouldn't we `#ifdef WIN32` this section?

Reply:
Oops. I just realized my mistake: I assumed the `#ifdef WIN32` in line 27 would span the entire file. But it doesn't.
The only part that keeps the function from being compiled on (our) other platforms is the `#if !defined(HAVE_SOCKETPAIR) && !defined(CURL_DISABLE_SOCKETPAIR)`. But there might be other platforms - that we don't use but are supported by libcurl - which don't `HAVE_SOCKETPAIR`.

Hm. That makes things more difficult. Is there an easy, portable way to wait for "ready to accept" without using `select`? Do you know if `poll` is available on all supported non-Windows systems?